drivers/boston_reports/app/models/boston_reports/community_of_origin.rb

```diff
@@ -0,0 +1,265 @@
+      { enrollment: { earliest_enrollment_location_history: :place } }
+    end
+
+    def across_the_country_data
+      @across_the_country_data ||= begin
+        percent_of_clients_data = enrolled_with_community_of_origin.group(:state).count.map do |state, count|
+          percentage = percent(numerator: count, denominator: count_enrolled_with_community_of_origin)
+          {
+            name: state,
+            count: count,
+            total: count_enrolled_with_community_of_origin,
+            percent: percentage,
+            display_percent: ActiveSupport::NumberHelper.number_to_percentage(percentage, precision: 1, strip_insignificant_zeros: true),
+          }
+        end
+        percent_names = percent_of_clients_data.map { |d| d[:name] }
+        GrdaWarehouse::Shape::State.where(name: percent_names).map do |state|
+          state.geo_json_properties.merge(percent_of_clients_data.detect { |d| d[:name] == state.name })
+        end.sort_by { |d| d[:percent] }.reverse
+      end
+    end
+
+    def across_my_state_data
+      @across_my_state_data ||= enrolled_with_community_of_origin_my_state.group('shape_cocs.cocnum').count.map do |coc_num, count|
+        percentage = percent(numerator: count, denominator: count_enrolled_with_community_of_origin)
+        {
+          name: HudUtility2024.coc_codes(coc_num),
+          count: count,
+          total: count_enrolled_with_community_of_origin,
+          percent: percentage,
+          display_percent: ActiveSupport::NumberHelper.number_to_percentage(percentage, precision: 1, strip_insignificant_zeros: true),
+        }
+      end.sort_by { |d| d[:percent] }.reverse
+    end
+
+    def my_state_data
+      across_the_country_data.detect { |d| d[:name] == GrdaWarehouse::Shape::State.my_state.first.name }
+    end
+
+    private def zip_code_scope
+      enrolled_with_community_of_origin.group(:zipcode).count.
+        sort_by(&:last).
+        reverse # Ensure descending order
+    end
+
+    def top_zip_codes_data
+      zip_code_scope.first(ZIP_LIMIT).map do |zip, count|
+        percentage = percent(numerator: count, denominator: count_enrolled_with_community_of_origin)
+        {
+          zip_code: zip,
+          count: count,
+          total: count_enrolled_with_community_of_origin,
+          percent: percentage,
+          display_percent: ActiveSupport::NumberHelper.number_to_percentage(percentage, precision: 1, strip_insignificant_zeros: true),
+        }
+      end
+    end
+
+    def zip_code_shape_data
+      zips = zip_code_scope.first(ZIP_LIMIT).map(&:first)
+      GrdaWarehouse::Shape.geo_collection_hash(GrdaWarehouse::Shape::ZipCode.where(zcta5ce10: zips))
+    end
+
+    def zip_code_colors
+      [
+        { color: '#BF216B', range: [0.02] },
+        { color: '#F22797', range: [0.02, 0.05] },
+        { color: '#F2BC1B', range: [0.05, 0.1] },
+        { color: '#F26A1B', range: [0.1, 0.15] },
+        { color: '#F5380E', range: [0.15] },
```

Comment:
Maybe use Ruby ranges here? I was confused by `range: [0.02]` and maybe `(.. 0.02)` would be clearer?

Reply:
Thanks, leaving, but added note, this is used by JavaScript, converting to ranges would probably break that and require a JS re-write.